### PR TITLE
Add DeviceMetaData class.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Adds standalone DeviceMetaData class. First step in #4743 .

### CODE REVIEW

File: cirq-core/cirq/devices/device.py [hunk 1]
```diff
@@ -178,3 +179,83 @@ def __iter__(self) -> Iterator['cirq.Qid']:
...
+class DeviceMetaData:
+    """Parent type for all device specific metadata classes."""
+
+    def qubit_set(self) -> Optional[FrozenSet['cirq.Qid']]:
+        """Returns a set of qubits on the device, if possible.
+
+        Returns:
+            Frozenset of qubits on device if specified, otherwise None.
+        """
+        return cast(Optional[FrozenSet['cirq.Qid']], self._qubits_set)
```
Comment (95-martin-orion) [inline cirq-core/cirq/devices/device.py:194]: `cast` shouldn't be necessary here - you can define the type of `self._qubits_set` in the constructor.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/devices/device.py [hunk 2]
```diff
@@ -178,3 +179,83 @@ def __iter__(self) -> Iterator['cirq.Qid']:
...
+
+    def nx_graph(self) -> Optional['nx.Graph']:
+        """Returns a nx.Graph where nodes are qubits are couple-able qubits.
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply cirq-core/cirq/devices/device.py]: Done.

File: cirq-core/cirq/devices/device.py [hunk 3]
```diff
@@ -178,3 +179,83 @@ def __iter__(self) -> Iterator['cirq.Qid']:
...
+    @classmethod
+    def _from_json_dict_(cls, qubits, nx_graph, obj_kwargs, **kwargs):
+        if qubits == '':
+            qubits = None
+        graph_obj = None
+        if nx_graph != '':
+            graph_obj = nx.readwrite.json_graph.node_link_graph(nx_graph)
+        return cls(qubits, graph_obj, **obj_kwargs)
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply cirq-core/cirq/devices/device.py]: Done.

File: cirq-core/cirq/devices/device.py [hunk 4]
```diff
@@ -178,3 +179,83 @@ def __iter__(self) -> Iterator['cirq.Qid']:
 
     def __contains__(self, item: 'cirq.Qid') -> bool:
         return item in self.qids
+
+
+@value.value_equality
+class DeviceMetaData:
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply cirq-core/cirq/devices/device.py]: Fixed.

File: cirq-core/cirq/devices/device.py (hunk 1 above)
Comment (MichaelBroughton) [reply cirq-core/cirq/devices/device.py]: Changed.

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**